### PR TITLE
Make utilities work for the newer version of JAX

### LINKED ISSUE #112: Some custom primitives does not work with the newest version of jax
The error happens because new version of jax requires a specific pattern for using `jit`'s static_argnums. See https://github.com/google/jax/pull/595 for more context. The failed functions include:

- [ ] xlogy
- [ ] xlog1py
- [ ] standard_gamma

In addition, I think that we can simplify the implementation of these functions by using decorator `custom_transform` as in `cumsum`, `cumprod`.

cc @neerajprad

### PR DESCRIPTION
**Pull request by fehiepsi**
Fix #112 by using `custom_transform` decorator.

In addition, this PR fixes tests with for fastmath mode (Peter (jax devs) mentioned that `np.inf`, `np.nan`, `np.exp` is fixed in the newer version of jaxlib so I would like to test it out and fixes errors during the way).

### CODE REVIEW

File: numpyro/distributions/util.py [hunk 1]
```diff
@@ -395,6 +282,103 @@ def multinomial_rvs(key, n, p, shape=()):
     return np.reshape(samples_2D, shape + p.shape[-1:]) - excess
 
 
+def multinomial(key, p, n, shape=()):
+    return _multinomial(key, p, n, shape)
+
+
+def _xlogy_jvp_lhs(g, x, y):
```
Comment (fehiepsi) [inline numpyro/distributions/util.py]: I group all samplers to one place, so other utilities are grouped into one place too.

File: numpyro/distributions/util.py [hunk 2]
```diff
@@ -361,23 +244,27 @@ def body_fn(val):
...
     return lax.scatter_add(operand, indices, updates,
                            lax.ScatterDimensionNumbers(update_window_dims=(),
                                                        inserted_window_dims=(0,),
                                                        scatter_dims_to_operand_dims=(0,)))
 
 
-@partial(jit, static_argnums=(1, 3))
-def multinomial_rvs(key, n, p, shape=()):
+@partial(jit, static_argnums=(2, 3))
+def _multinomial(key, p, n, shape=()):
```
Comment (fehiepsi) [inline numpyro/distributions/util.py]: the pattern
```
@partial(jit, static_argnums=(2, 3))
def _f(a,b,c,d):
    ...

def f(a,b,c,d):
    return _f(a,b,c,d)
```
is used in `jax.random` to support different ways to set `args`, `kwargs` to samplers. Without this pattern, calling `f(a,b,c=c,d=d)` will trigger an error that `c` and `d` must be args, not kwargs.

File: numpyro/distributions/util.py [hunk 3]
```diff
@@ -395,6 +282,103 @@ def multinomial_rvs(key, n, p, shape=()):
...
 
 
+def multinomial(key, p, n, shape=()):
+    return _multinomial(key, p, n, shape)
+
+
+def _xlogy_jvp_lhs(g, x, y):
+    shape = lax.broadcast_shapes(np.shape(g), np.shape(y))
+    g = np.broadcast_to(g, shape)
+    y = np.broadcast_to(y, shape)
```
Comment (fehiepsi) [inline numpyro/distributions/util.py]: `lax._brcast` is not exposed (changed to `lax.lax._brcast`) in the newer version of JAX, so I use `np.broadcast_to` instead

File: numpyro/distributions/util.py (hunk 2 above)
Comment (neerajprad) [reply numpyro/distributions/util.py]: Hmm..what's the reason for this? Is this related to the change that kwargs are not treated as static args by default in jit?

File: test/test_distributions.py [hunk 4]
```diff
@@ -242,6 +242,14 @@ def test_log_prob(jax_dist, sp_dist, params, prepend_shape, jit):
         expected = sp_dist.logpdf(samples)
     except AttributeError:
         expected = sp_dist.logpmf(samples)
+    except ValueError as e:
+        # precision issue: np.sum(x / np.sum(x)) = 0.99999994 != 1
```
Comment (neerajprad) [inline test/test_distributions.py]: Maybe we should lower this precision value in the transform or have it be dependent on the data, since I think scipy would do the normalization behind the scenes anyways.

File: numpyro/distributions/util.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/distributions/util.py]: Yes, kwargs are not allowed in the positions of static args in the newer version (you can see that commit [here](https://github.com/google/jax/pull/595)).

File: test/test_distributions.py (hunk 4 above)
Comment (fehiepsi) [reply test/test_distributions.py]: This is indeed scipy problem. Scipy works in float64 so their check is `< 1e-10` would be fine for the output of `np.sum(x / np.sum(x))`. Our check `1e-6` in float32 should be fine but our samples would not be valid samples for scipy.
